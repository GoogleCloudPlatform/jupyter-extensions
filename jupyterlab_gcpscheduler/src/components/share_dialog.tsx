import { LearnMoreLink, COLORS } from 'gcp_jupyterlab_shared';
import * as React from 'react';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import { classes, stylesheet } from 'typestyle';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';

interface Props {
  cloudBucket: string;
  shareLink: string;
  handleClose?: () => void;
}

interface State {
  openDialog: boolean;
  openSnackbar: boolean;
}

const localStyles = stylesheet({
  text: {
    fontSize: '13px !important',
    color: COLORS.base + ' !important',
  },
  fullWidth: {
    width: '100%',
  },
  inlineStart: {
    paddingInlineStart: '16px',
    paddingTop: '16px',
    paddingBottom: '16px',
  },
  actionButtons: {
    margin: '12px 24px 24px 24px',
  },
  primaryButton: {
    color: COLORS.focus + '!important',
  },
});

export class ShareDialog extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { openDialog: false, openSnackbar: false };
    this.handleClickOpen = this.handleClickOpen.bind(this);
    this.handleCopy = this.handleCopy.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.handleSnackbarClose = this.handleSnackbarClose.bind(this);
  }

  handleClickOpen() {
    this.setState({ openDialog: true });
  }

  handleCopy() {
    navigator.clipboard.writeText(this.props.shareLink);
    this.setState({ openSnackbar: true });
  }

  handleClose() {
    this.setState({ openDialog: false, openSnackbar: false });
    if (this.props.handleClose) {
      this.props.handleClose();
    }
  }

  handleSnackbarClose() {
    this.setState({ openSnackbar: false });
  }

  render() {
    return (
      <React.Fragment>
        <p onClick={this.handleClickOpen} className={localStyles.fullWidth}>
          Share execution result
        </p>
        <Dialog
          open={this.state.openDialog}
          onClose={this.handleClose}
          aria-labelledby="share-dialog-title"
          aria-describedby="share-dialog-description"
          maxWidth="sm"
          fullWidth={true}
        >
          <DialogTitle id="share-dialog-title">
            Share execution results in this project
          </DialogTitle>
          <DialogContent>
            <DialogContentText
              className={localStyles.text}
              id="share-dialog-description"
            >
              To share this execution result with people or groups, grant view
              permissions to the Cloud Storage bucket containing this result. If
              this has been done before, skip steps 2 through 4.
            </DialogContentText>
            <ol className={classes(localStyles.inlineStart, localStyles.text)}>
              <li>Click the "COPY LINK" button below</li>
              <li>
                Go to the{' '}
                <LearnMoreLink
                  href={this.props.cloudBucket}
                  text="results bucket"
                />
                .
              </li>
              <li>Click "Add"</li>
              <li>
                Enter one or more emails then select the "Storage Object Viewer"
                role
              </li>
            </ol>
            <DialogContentText
              className={localStyles.text}
              id="share-dialog-description"
            >
              You can now share links to all execution results stored in this
              bucket, including the one you copied, with the people or groups
              granted the view permission.
            </DialogContentText>
          </DialogContent>
          <DialogActions className={localStyles.actionButtons}>
            <Button
              className={localStyles.primaryButton}
              onClick={this.handleClose}
              color="primary"
            >
              Cancel
            </Button>
            <Button
              className={localStyles.primaryButton}
              onClick={this.handleCopy}
              color="primary"
              autoFocus
            >
              Copy Link
            </Button>
          </DialogActions>
        </Dialog>
        <Snackbar
          open={this.state.openSnackbar}
          onClose={this.handleSnackbarClose}
          message="Copied to clipboard"
        />
      </React.Fragment>
    );
  }
}
