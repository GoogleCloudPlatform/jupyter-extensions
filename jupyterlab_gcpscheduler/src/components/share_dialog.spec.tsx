/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/no-unused-vars */
import { shallow } from 'enzyme';
import * as React from 'react';
import { ShareDialog } from './share_dialog';

Object.assign(navigator, {
  clipboard: {
    writeText: jest.fn(),
  },
});

describe('ShareDialog', () => {
  const mockOnHandleClose = jest.fn();
  jest.spyOn(navigator.clipboard, 'writeText');
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Renders closed', () => {
    const component = shallow(
      <ShareDialog
        cloudBucket="http://storage.bucket"
        shareLink="http://share.link"
        handleClose={mockOnHandleClose}
      />
    );
    expect(component).toMatchSnapshot();
    expect(component.state('openDialog')).toBe(false);
  });

  it('Renders open', () => {
    const component = shallow(
      <ShareDialog
        cloudBucket="http://storage.bucket"
        shareLink="http://share.link"
        handleClose={mockOnHandleClose}
      />
    );
    component.find('p').simulate('click');
    expect(component).toMatchSnapshot();
    expect(component.state('openDialog')).toBe(true);
  });

  it('Handles copy', () => {
    const component = shallow(
      <ShareDialog
        cloudBucket="http://storage.bucket"
        shareLink="http://share.link"
        handleClose={mockOnHandleClose}
      />
    );
    component.find('p').simulate('click');
    component
      .find('WithStyles(ForwardRef(Button))')
      .at(1)
      .simulate('click');
    expect(navigator.clipboard.writeText).toHaveBeenCalledWith(
      'http://share.link'
    );
    expect(component.state('openSnackbar')).toBe(true);
  });

  it('Handles cancel', () => {
    const component = shallow(
      <ShareDialog
        cloudBucket="http://storage.bucket"
        shareLink="http://share.link"
        handleClose={mockOnHandleClose}
      />
    );
    component.find('p').simulate('click');
    component
      .find('WithStyles(ForwardRef(Button))')
      .at(0)
      .simulate('click');
    expect(component.state('openSnackbar')).toBe(false);
    expect(component.state('openDialog')).toBe(false);
    expect(mockOnHandleClose).toHaveBeenCalled();
  });
});
