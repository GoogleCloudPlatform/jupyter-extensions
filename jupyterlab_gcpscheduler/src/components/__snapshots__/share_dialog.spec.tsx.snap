// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareDialog Renders closed 1`] = `
<Fragment>
  <p
    className="fullWidth_f96or0t"
    onClick={[Function]}
  >
    Share execution result
  </p>
  <WithStyles(ForwardRef(Dialog))
    aria-describedby="share-dialog-description"
    aria-labelledby="share-dialog-title"
    fullWidth={true}
    maxWidth="sm"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="share-dialog-title"
    >
      Share execution results in this project
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))
        className="text_f1v1xbpa"
        id="share-dialog-description"
      >
        To share this execution result with people or groups, grant view permissions to the Cloud Storage bucket containing this result. If this has been done before, skip steps 2 through 4.
      </WithStyles(ForwardRef(DialogContentText))>
      <ol
        className="inlineStart_fm4i5kc text_f1v1xbpa"
      >
        <li>
          Click the "COPY LINK" button below
        </li>
        <li>
          Go to the
           
          <LearnMoreLink
            href="http://storage.bucket"
            text="results bucket"
          />
          .
        </li>
        <li>
          Click "Add"
        </li>
        <li>
          Enter one or more emails then select the "Storage Object Viewer" role
        </li>
      </ol>
      <WithStyles(ForwardRef(DialogContentText))
        className="text_f1v1xbpa"
        id="share-dialog-description"
      >
        You can now share links to all execution results stored in this bucket, including the one you copied, with the people or groups granted the view permission.
      </WithStyles(ForwardRef(DialogContentText))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))
      className="actionButtons_ft2surt"
    >
      <WithStyles(ForwardRef(Button))
        className="primaryButton_f18dymgn"
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        autoFocus={true}
        className="primaryButton_f18dymgn"
        color="primary"
        onClick={[Function]}
      >
        Copy Link
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
  <WithStyles(ForwardRef(Snackbar))
    message="Copied to clipboard"
    onClose={[Function]}
    open={false}
  />
</Fragment>
`;

exports[`ShareDialog Renders open 1`] = `
<Fragment>
  <p
    className="fullWidth_f96or0t"
    onClick={[Function]}
  >
    Share execution result
  </p>
  <WithStyles(ForwardRef(Dialog))
    aria-describedby="share-dialog-description"
    aria-labelledby="share-dialog-title"
    fullWidth={true}
    maxWidth="sm"
    onClose={[Function]}
    open={true}
  >
    <WithStyles(ForwardRef(DialogTitle))
      id="share-dialog-title"
    >
      Share execution results in this project
    </WithStyles(ForwardRef(DialogTitle))>
    <WithStyles(ForwardRef(DialogContent))>
      <WithStyles(ForwardRef(DialogContentText))
        className="text_f1v1xbpa"
        id="share-dialog-description"
      >
        To share this execution result with people or groups, grant view permissions to the Cloud Storage bucket containing this result. If this has been done before, skip steps 2 through 4.
      </WithStyles(ForwardRef(DialogContentText))>
      <ol
        className="inlineStart_fm4i5kc text_f1v1xbpa"
      >
        <li>
          Click the "COPY LINK" button below
        </li>
        <li>
          Go to the
           
          <LearnMoreLink
            href="http://storage.bucket"
            text="results bucket"
          />
          .
        </li>
        <li>
          Click "Add"
        </li>
        <li>
          Enter one or more emails then select the "Storage Object Viewer" role
        </li>
      </ol>
      <WithStyles(ForwardRef(DialogContentText))
        className="text_f1v1xbpa"
        id="share-dialog-description"
      >
        You can now share links to all execution results stored in this bucket, including the one you copied, with the people or groups granted the view permission.
      </WithStyles(ForwardRef(DialogContentText))>
    </WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogActions))
      className="actionButtons_ft2surt"
    >
      <WithStyles(ForwardRef(Button))
        className="primaryButton_f18dymgn"
        color="primary"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        autoFocus={true}
        className="primaryButton_f18dymgn"
        color="primary"
        onClick={[Function]}
      >
        Copy Link
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(DialogActions))>
  </WithStyles(ForwardRef(Dialog))>
  <WithStyles(ForwardRef(Snackbar))
    message="Copied to clipboard"
    onClose={[Function]}
    open={false}
  />
</Fragment>
`;
